File: packages/server-flask/open_passkey_flask/stores.py
"""Challenge and credential store interfaces with in-memory implementations."""

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from threading import Lock
from time import time


class PasskeyError(Exception):
    def __init__(self, message: str, status_code: int = 400):
        super().__init__(message)
        self.status_code = status_code


@dataclass
class StoredCredential:
    credential_id: bytes
    public_key_cose: bytes
    sign_count: int
    user_id: str
    prf_salt: bytes | None = None
    prf_supported: bool = False


class ChallengeStore(ABC):
    """Manages single-use, time-limited WebAuthn challenges. Must be thread-safe."""

    @abstractmethod
    def store(self, key: str, challenge: str, timeout_seconds: float) -> None: ...

    @abstractmethod
    def consume(self, key: str) -> str:
        """Retrieve and delete a challenge. Raises PasskeyError if not found/expired."""
        ...


class CredentialStore(ABC):
    """Manages WebAuthn credential persistence. Must be thread-safe."""

    @abstractmethod
    def store(self, cred: StoredCredential) -> None: ...

    @abstractmethod
    def get(self, credential_id: bytes) -> StoredCredential: ...

    @abstractmethod
    def get_by_user(self, user_id: str) -> list[StoredCredential]: ...

    @abstractmethod
    def update(self, cred: StoredCredential) -> None: ...

    @abstractmethod
    def delete(self, credential_id: bytes) -> None: ...


# --- In-memory implementations ---

_CLEANUP_INTERVAL = 100


@dataclass
class _ChallengeEntry:
    challenge: str
    expires_at: float


class MemoryChallengeStore(ChallengeStore):
    def __init__(self):
        self._lock = Lock()
        self._entries: dict[str, _ChallengeEntry] = {}
        self._write_count = 0

    def store(self, key: str, challenge: str, timeout_seconds: float) -> None:
        with self._lock:
            self._entries[key] = _ChallengeEntry(challenge, time() + timeout_seconds)
            self._write_count += 1
            if self._write_count >= _CLEANUP_INTERVAL:
                self._write_count = 0
                now = time()
                self._entries = {
                    k: v for k, v in self._entries.items() if v.expires_at > now
                }

    def consume(self, key: str) -> str:
        with self._lock:
            entry = self._entries.pop(key, None)
            if entry is None or time() > entry.expires_at:
                raise PasskeyError("challenge not found or expired")
            return entry.challenge


class MemoryCredentialStore(CredentialStore):
    def __init__(self):
        self._lock = Lock()
        self._creds: list[StoredCredential] = []

    def store(self, cred: StoredCredential) -> None:
        with self._lock:
            self._creds.append(cred)

    def get(self, credential_id: bytes) -> StoredCredential:
        with self._lock:
            for c in self._creds:
                if c.credential_id == credential_id:
                    return c
        raise PasskeyError("credential not found")

    def get_by_user(self, user_id: str) -> list[StoredCredential]:
        with self._lock:
            return [c for c in self._creds if c.user_id == user_id]

    def update(self, cred: StoredCredential) -> None:
        with self._lock:
            for i, c in enumerate(self._creds):
                if c.credential_id == cred.credential_id:
                    self._creds[i] = cred
                    return
        raise PasskeyError("credential not found")

    def delete(self, credential_id: bytes) -> None:
        with self._lock:
            for i, c in enumerate(self._creds):
                if c.credential_id == credential_id:
                    self._creds.pop(i)
                    return
        raise PasskeyError("credential not found")

File: packages/server-spring/src/main/java/com/openpasskey/spring/Stores.java
package com.openpasskey.spring;

import java.time.Instant;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Challenge and credential store interfaces with thread-safe in-memory implementations.
 */
public class Stores {

    // --- Stored credential ---

    public record StoredCredential(
        byte[] credentialId,
        byte[] publicKeyCose,
        long signCount,
        String userId,
        byte[] prfSalt,
        boolean prfSupported
    ) {
        public StoredCredential withSignCount(long newSignCount) {
            return new StoredCredential(credentialId, publicKeyCose, newSignCount, userId, prfSalt, prfSupported);
        }
    }

    // --- Challenge store interface ---

    public interface ChallengeStore {
        void store(String key, String challenge, long timeoutSeconds);
        String consume(String key) throws PasskeyException;
    }

    // --- Credential store interface ---

    public interface CredentialStore {
        void store(StoredCredential cred);
        StoredCredential get(byte[] credentialId) throws PasskeyException;
        List<StoredCredential> getByUser(String userId);
        void update(StoredCredential cred) throws PasskeyException;
        void delete(byte[] credentialId) throws PasskeyException;
    }

    // --- PasskeyException ---

    public static class PasskeyException extends Exception {
        private final int statusCode;
        public PasskeyException(String message) { this(message, 400); }
        public PasskeyException(String message, int statusCode) {
            super(message);
            this.statusCode = statusCode;
        }
        public int getStatusCode() { return statusCode; }
    }

    // --- In-memory challenge store ---

    private record ChallengeEntry(String challenge, Instant expiresAt) {}

    public static class MemoryChallengeStore implements ChallengeStore {
        private static final int CLEANUP_INTERVAL = 100;
        private final ConcurrentHashMap<String, ChallengeEntry> entries = new ConcurrentHashMap<>();
        private final AtomicInteger writeCount = new AtomicInteger(0);

        @Override
        public void store(String key, String challenge, long timeoutSeconds) {
            entries.put(key, new ChallengeEntry(challenge, Instant.now().plusSeconds(timeoutSeconds)));
            if (writeCount.incrementAndGet() >= CLEANUP_INTERVAL) {
                writeCount.set(0);
                Instant now = Instant.now();
                entries.entrySet().removeIf(e -> now.isAfter(e.getValue().expiresAt()));
            }
        }

        @Override
        public String consume(String key) throws PasskeyException {
            ChallengeEntry entry = entries.remove(key);
            if (entry == null || Instant.now().isAfter(entry.expiresAt())) {
                throw new PasskeyException("challenge not found or expired");
            }
            return entry.challenge();
        }
    }

    // --- In-memory credential store ---

    public static class MemoryCredentialStore implements CredentialStore {
        private final CopyOnWriteArrayList<StoredCredential> creds = new CopyOnWriteArrayList<>();

        @Override
        public void store(StoredCredential cred) {
            creds.add(cred);
        }

        @Override
        public StoredCredential get(byte[] credentialId) throws PasskeyException {
            for (StoredCredential c : creds) {
                if (Arrays.equals(c.credentialId(), credentialId)) return c;
            }
            throw new PasskeyException("credential not found");
        }

        @Override
        public List<StoredCredential> getByUser(String userId) {
            List<StoredCredential> result = new ArrayList<>();
            for (StoredCredential c : creds) {
                if (c.userId().equals(userId)) result.add(c);
            }
            return result;
        }

        @Override
        public void update(StoredCredential cred) throws PasskeyException {
            for (int i = 0; i < creds.size(); i++) {
                if (Arrays.equals(creds.get(i).credentialId(), cred.credentialId())) {
                    creds.set(i, cred);
                    return;
                }
            }
            throw new PasskeyException("credential not found");
        }

        @Override
        public void delete(byte[] credentialId) throws PasskeyException {
            boolean removed = creds.removeIf(c -> Arrays.equals(c.credentialId(), credentialId));
            if (!removed) throw new PasskeyException("credential not found");
        }
    }
}
